src/go/rpk/pkg/config/params.go

```diff
@@ -552,3 +568,133 @@ func (c *Config) addUnsetDefaults() {
 		r.AdminAPI.Addresses = []string{"127.0.0.1:9644"}
 	}
 }
+
+// Set allow to set a single configuration property by passing a key value pair
+//
+//   Key:    string containing the yaml property tag, e.g: 'rpk.admin_api'.
+//   Value:  string representation of the value, either single value or partial
+//           representation if a format is passed (yaml / json).
+//   Format: either single, json, or yaml (default: single).
+func (c *Config) Set(key, value, format string) error {
+	if key == "" {
+		return fmt.Errorf("key field must not be empty")
+	}
+	props := strings.Split(key, ".")
+	rv := reflect.ValueOf(c).Elem()
+	found, other, err := getField(props, rv)
+	if err != nil {
+		return err
+	}
+	isOther := other != reflect.Value{}
+
+	field := found
+	if isOther {
+		field = other
+	}
+
+	if field.CanAddr() {
+		i := field.Addr().Interface()
+		in := value
+		switch strings.ToLower(format) {
+		case "yaml", "single", "":
+			if isOther {
+				p := props[len(props)-1]
+				in = fmt.Sprintf("%s: %s", p, value)
+			}
+			err = yaml.Unmarshal([]byte(in), i)
```

Comment:
Why doesn't this properly call weak unmarshaling for socket addresses, etc? I think you brought up that setting redpanda.seed_servers is different than setting `redpanda` with a value containing seed servers?

Reply:
It doesn't call weak unmarshaling for any of the `one_or_many` weak types, since we created other types for them:

```
[]NamedSocketAddress -> NamedSocketAddresses
[]SeedServer -> SeedServers
... 
```

Reply:
I have an idea for how to solve this for the yaml case, but not the json. In the yaml case, when we return, we could have a huge type switch:
```
case []string => return weakStrings(xyz)
case []NamedSocketAddress => return namedSocketAddresses(xyz)
```

this only works for yaml because weak decoding is only implemented for yaml
this theoretically could be supported for json in a followup PR if we want,
but I think this should be tracked as a separate issue: allow setting slices with one element in set

Reply:
Created: https://github.com/redpanda-data/redpanda/issues/5264